contributors/design-proposals/storage/volume-topology-scheduling.md

```diff
@@ -266,41 +304,329 @@ const (
+label for each topology key specified and one of its allowed values. For
+example, a topology specified with "zone: [us-central1-a, us-central1-b]" means
+that the node must have the label "zone: us-central1-a" or "zone:
+us-central1-b".
+
+A StorageClass update controller can be provided to convert existing topology
+restrictions specified under plugin-specific parameters to the new AllowedTopology
+field.
+
+The AllowedTopology will also be provided to provisioners as a new field, detailed in
+the provisioner section.  Provisioners can use the allowed topology information
+in the following scenarios:
+* StorageClass is using the default immediate binding mode. This is the
+  legacy topology-unaware behavior. In this scenario, the volume could be
+  provisioned in a domain that cannot run the Pod since it doesn't take any
+  scheduler input.
+* For volumes that span multiple domains, the AllowedTopology can restrict those
+  additional domains.  However, special care must be taken to avoid specifying
+  conflicting topology constraints in the Pod. For example, the administrator could
+  restrict a multi-zonal volume to zones 'zone1' and 'zone2', but the Pod could have
+  constraints that restrict it to 'zone1' and 'zone3'.  If 'zone1'
+  fails, the Pod cannot be scheduled to the intended failover zone.
+
+Kubernetes will leave validation and enforcement of the AllowedTopology content up
+to the provisioner.
+
+TODO: would this be better represented as part of StorageClass quota since that
+is where admins currently specify policies?
+However, ResourceQuota currently only supports quantity, and topology
+restrictions have to be evaluated during scheduling, not admission.
+
+#### Zonal Example
+This example restricts the volumes provisioned to zones us-central1-a and
+us-central1-b.
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  Name: zonal-class
+provisioner: kubernetes.io/gce-pd
+parameters:
+  type: pd-standard
+allowedTopology:
+  required:
+    failure-domain.beta.kubernetes.io/zone:
+      values:
+      - us-central1-a
+      - us-central1-b
+```
+
+#### Multi-Zonal Example
+This example restricts the volume's primary and failover zones
+to us-central1-a and us-central1-b. Topologies that are incompatible with the
+storage provider parameters will be enforced by the provisioner.
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  Name: multi-zonal-class
+provisioner: kubernetes.io/gce-pd
+parameters:
+  type: pd-standard
+  replicas: 2
+allowedTopology:
+  required:
+    failure-domain.beta.kubernetes.io/zone:
+      values:
+      - us-central1-a
+      - us-central1-b
+```
+
+### Storage Provider Capacity Reporting
+Some volume types have capacity limitations per topology domain. For example,
+local volumes have limited capacity per node. Provisioners for these volume
+types can report these capacity limitations through a new StorageClass status
+field.
```

Comment:
I've been warned before against having constantly changing metrics (like resource usage) be placed as an API status, and one argument is that it could easily overwhelm the API server. Maybe the capacity information should be plumbed to the scheduler through existing resource usage monitoring pipelines.

Might be good to get a review from sig-instrumentation?

Reply:
The capacity reported by the provisioner should be total capacity, not available capacity.  I don't anticipate total capacity changing very frequently.

The challenge with existing resource usage tracking is that it is done on a per node level, whereas volume resource limits may not necessarily be at the node boundary.

Reply:
There are plans to eventually plumb PVC usage stats through Metrics Server, as part of the work to get these stats exposed to kubectl. I believe the scheduler also talks to the Metrics Server.

Currently there's a limitation that PVC usage stats can only be measured when the underlying volume is mounted in a pod, but maybe we can expand this feature and add the capability of having volume plugins report these stats as well.

@jingxu97

Reply:
I believe for PVC usage stats, it's still tied to a Pod that is scheduled to a specific node.  That way you can hit all the kubelet metrics endpoints.  For this design to handle getting capacity from metrics, we would need to collect all the metrics from a central entity, and not per node to handle volumes with topologies spanning more than a single node.

I am unaware of the scheduler talking to the metrics server.  My understanding is it uses resource capacity reported in the Node object.  @bsalamat could you clarify?

But besides the initial StorageClass creation, I don't anticipate reported capacity to be updated very frequently.

Reply:
Can we find the details of the feature somewhere?

Reply:
Here, we have also have the challenge that external provisioners need to report their capacity somewhere, whether on an API object (like proposed), or to some central server.

Reply:
Yeah, as I'm also focusing on design of local storage provisioning, I wonder what the way to talk to the metrics server looks like.

I guess users are more likely to look forward to things they are familiar

Reply:
I discussed a bit with @verult, and we agreed that going the metrics server approach would require a lot of extra plumbing that doesn't exist today.  So we're still leaning towards the current approach of reporting total (not used) capacity in the StorageClass and calculating and caching used from provisioned PVCs.